solarforecastarbiter/datamodel.py

```diff
@@ -934,6 +938,227 @@ def __post_init__(self):
+    net : bool
+        If True, compute the 'net' aggregate error instead of first calcuating
+        the absolute error before performing the aggregation.
+    fill : str
+        Fill method to apply for times between those specified in `times`.
+        Options are 'forward' or 'backward'.
+    timezone : str, default None
+        IANA timezone string to use when constructing datetimes. If None,
+        the timezone of the observations is used, which is the report
+        timezone when calculated in a report.
+    """
+    times: Tuple[datetime.time, ...]
+    costs: Tuple[float, ...]
+    aggregation: str
+    net: bool
+    fill: str
+    timezone: str = None
+    __post_init__ = __validate_cost__('times')
+
+
+@dataclass(frozen=True)
+class DatetimeCost(BaseModel):
+    """Cost values based on datetimes.
+
+    Parameters
+    ----------
+    datetimes : tuple of datetime.datetime
+       The datetimes to associate with each cost value
+    costs : tuple of float
+       The cost per unit error of the forecasted variable for each datetime.
+       Must have the same length as `datetimes`.
+    aggregation : str
+        Aggregation method to use after calculating cost for the error series.
+        Currently only 'sum' or 'mean' are available.
+    net : bool
+        If True, compute the 'net' aggregate error instead of first calcuating
+        the absolute error before performing the aggregation.
+    fill : str
+       Fill method to apply for datetimes between those specified in
+       `datetimes`. Options are 'forward' or 'backward'.
+    timezone : str, default None
+        IANA timezone string to use when constructing datetimes. If None,
+        the timezone of the observations is used, which is the report
+        timezone when calculated in a report.    """
+    datetimes: Tuple[pd.Timestamp, ...]
+    costs: Tuple[float, ...]
+    aggregation: str
+    net: bool
+    fill: str
+    timezone: str = None
+    __post_init__ = __validate_cost__('datetimes')
+
+
+@dataclass(frozen=True)
+class ConstantCost(BaseModel):
+    """A constant cost per unit error of the forecasted variable
+
+    Parameters
+    ----------
+    cost : float
+    aggregation : str
+        Aggregation method to use after calculating cost for the error series.
+        Currently only 'sum' or 'mean' are available.
+    net : bool
+        If True, compute the 'net' aggregate error instead of first calcuating
+        the absolute error before performing the aggregation.
+    """
+    cost: float
+    aggregation: str
+    net: bool
+    __post_init__ = __validate_cost__(None)
+
+
+@dataclass(frozen=True)
+class CostBand(BaseModel):
+    """Cost specification for one error band
+
+    Parameters
+    ----------
+    error_range : tuple(float, float)
+        Bounds of the error to apply the specified cost function to.
+        Inf and -Inf are valid range points, and the error may be positive or
+        negative.
```

Comment:
It depends on the ordering in ErrorBandCost. Basically, the bands are evaluated in order, so the first band gets inclusion of both endpoints. I think this precludes using np.digitize without lots of massaging instead of the masking function. I personally think this approach is easier to understand than including some additional parameter to define the bounding. The other "benefit" is that you don't need to coordinate the ranges to make sure there is no overlap. For example, the first range can be (-2, 2) inclusive, and the second band could specify (-inf, inf). So the first band would be used for [-2, 2] and the second for (-inf, -2) and (2, inf). That's what the not below is about and I will try and make this more clear in a separate documentation page.